File: MinPathTriangle.java
// Time Complexity : O(n*m)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

//we shall take a dp array which will keep track of path sum, of length triangle
//populate dp array with last row of triangle as we move step by step from bottom upward
//at each level i, for each j, take min value between that index and the next from the row below , 
//add it with current index from triangle and update the dp array
//after all levels are traversed, the value at dp[0] wpuld be our answer 

class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle==null||triangle.size()==0||triangle.get(0).size()==0){
            return 0;
        }
        
        int[] dp=new int[triangle.get(triangle.size()-1).size()];
        for(int i=0;i<dp.length;i++){
            dp[i]=triangle.get(dp.length-1).get(i);
        }
        for(int i=triangle.size()-2;i>=0;i--){
            for(int j=0;j<=i;j++){
                dp[j]=Math.min(dp[j],dp[j+1])+triangle.get(i).get(j);
            }
        }
        return dp[0];
    }
}

File: min_path_triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        #O(size of triangle)
        #O(1)
        #bottom to top
        #keep last row untouched
        #start checking from 2nd last row all the way to top
        if not triangle:
            return 0
        rows=len(triangle)
        for i in range(rows-2,-1,-1):
            for j in range(len(triangle[i])):
                first=triangle[i+1][j]+triangle[i][j]
                second=triangle[i+1][j+1]+triangle[i][j]
                triangle[i][j]=min(first,second)
        return triangle[0][0]
        
        
                
                
        
    
                